README.md
# Internation

[![Version](https://img.shields.io/npm/v/reveal.js-internation)](#) [![Downloads](https://img.shields.io/npm/dt/reveal.js-internation)](https://github.com/Martinomagnifico/reveal.js-internation/archive/refs/heads/master.zip)

A plugin for [Reveal.js](https://revealjs.com), that can let the end user choose a language for the presentation. Creating additional languages is made easy by letting you tag elements that should be translated. Running the presentation in 'makejson'-mode will generate a full JSON file of the whole current presentation. You can then start translating this file into other languages.

You should probably actually know these other languages (or someone who does) to be able to do so :-)

[<img src="https://martinomagnifico.github.io/reveal.js-internation/screenshot.png" width="100%">](https://martinomagnifico.github.io/reveal.js-internation/demo.html)



## Installation

### Regular installation

Copy the internation folder to the plugins folder of the reveal.js folder, like this: `plugin/internation`.

### npm installation

This plugin is published to, and can be installed from, npm.

```console
npm install reveal.js-internation
```
The Internation plugin folder can then be referenced from `node_modules/reveal.js-internation/plugin/internation`


## Setup

### JavaScript

There are two JavaScript files for Internation, a regular one, `internation.js`, and a module one, `internation.esm.js`. You only need one of them:

#### Regular 

If you're not using ES modules, for example, to be able to run your presentation from the filesystem, you can add it like this:

```html
<script type="text/javascript" src="dist/reveal.js"></script>
<script src="plugin/internation/internation.js"></script>
<script>
	Reveal.initialize({
		// ...
		plugins: [ Internation ]
	});
</script>
```
#### As a module 
If you're using ES modules, you can add it like this:

```html
<script type="module">
	// This will need a server
	import Reveal from './dist/reveal.esm.js';
	import Internation from './plugin/internation/internation.esm.js';
	Reveal.initialize({
		// ...
		plugins: [ Internation ]
	});
</script>
```


## Configuration

There are a few options that you can change from the Reveal.js options. The values below are default and do not need to be set if they are not changed.

```javascript
Reveal.initialize({
	// ...
	internation: {
		locale: "en",
		localename: 'English',
		langattribute: "data-i18n",
		switchselector: ".langchooser",
		html: true,
		languages: {},
		debug: false,
		makejson: false
	},
	plugins: [ Internation ]
});
```

* **`locale`**: The shortened version of the language that the presentation is made in. 
* **`localename`**: The human-readable name of the language that the presentation is made in. 
* **`langattribute`**: The name of the language attribute. Use this to 'tag' your texts.
* **`switchselector`**: The selector for switching languages while running the presentation.
* **`html`**: Boolean to allow HTML inside the translation files and tags. Set to 'true' by default.
* **`languages`**: You can put whole dictionary objects here, instead of JSON files, if you *really* want to allow local file access.
* **`debug`**: Boolean to allow debugging. Set to 'false' by default.
* **`makejson`**: Boolean to set 'makejson' mode. Set to 'false' by default. See "Step 3" below.


## Step 1. Define the current language.

Internation has English as the default language. You can change it if you like in the options.


## Step 2. Start tagging your text

Give your slides (the ones in which you want to translate texts) specific ID's. If you have an item (like a menubar) across multiple slides, you need to give it an ID as well.

Add a data-attribute of data-i18n="something" to the text you want to translate. You can skip elements that are common in all languages.

```html
<section id="intro">
    <h1 data-i18n="title">Hello!</h1>
    <p data-i18n="paragraph1">There is a <a href="#">link</a> in the paragraph</p>
</section>
```
Internation reads HTML by default. If some text contains a span or link, it is better to take that whole piece for translation. If you turn the HTML setting off in the options (you can), you would need to make spans of every text that is not a link.


## Step 3. Run your presentation in 'makejson'-mode.

After having added your data-attributes in step 2, you can now turn on the 'makejson'-mode in the Internation options:  

```javascript
Reveal.initialize({
    //
    internation: {
        makejson: true
    },
    plugins: [ Internation ]
});
```

Next, and this is important, **load the page with this URL parameter: ...index.html?makejson**. This will immediately download a JSON file of your current HTML structure. The URL step is a safeguard for if you accidentally publish with makejson:true in the options turned on.

### Example dictionary/JSON

Here's an example of an English language file. This example has a menubar that is visible on all slides: 

```javascript
{
    "slides" : {
        "intro":{
            "title":"Hi there!",
            "paragraph1":"This is a paragraph"
        },
        "animals":{
            "title":"Animals",
            "dog":"Dog",
            "cat":"Cat",
        }
    },
    "menubar" : {
        "title":"Here is the title",
    }
}
```

## Step 4. Start translating

Make a copy of your first generated JSON. Give the file a name that is obvious (so, fr.json for French).
**Now start translating!** When you are finished, put the file somewhere in your Reveal presentation folder.

## Step 5A. Refer to your new language file(s)…

A language in Internation contains both a name, a dictionary and an optional direction. This example shows that a JSON file is used for the dictionary. JSON files will only work if there is a server. It will not work with the file:// protocol. 

The direction can be added if you want to change the HTML reading direction when you change the language to any 'rtl' language like Hebrew or Arabic.

```javascript
Reveal.initialize({
    //
    internation: {
        languages: {
            fr: {
                name: "Français",
                dictionary: "fr.json"
            },
            ar: {
                name: "العربية",
                dictionary: "ar.json",
                direction: "rtl"
            }
        }
    },
    plugins: [ Internation ]
});
```

## Step 5B. …or language object

If you *really* want to allow local file access, you can put the whole dictionary object in the specific language object in your options:

```javascript
Reveal.initialize({
    //
    internation: {
        languages: {
            fr: {
                name: "Français",
                dictionary: {
                    "slides" : {
                        "intro":{"title":"Bonjour!"}
                    }
                }
            }
        }
    },
    plugins: [ Internation ]
});
```

## Switching languages

If you want multiple languages, you will probably also want users to choose. Internation listens to 'change' events of either a select or a set of radio buttons inside the presentation. By default, this is any element with the class `langchooser`, but that classname can be changed in the options.

You can put these on a slide (like the first page of the demo with the flags) or outside them in an element that is visible on multiple slides (like the menubar in the examples). 

The demo has a fake select in the menubar that works with radio buttons. You can pick and style your own way of switching.



## Like it?
If you like it, please star this repo! 

And if you want to show off what you made with it, please do :-)


## License
MIT licensed

Copyright (C) 2022 Martijn De Jongh (Martino)


CHANGELOG.md
# Changelog

## [1.1.0] - 2022-11-13
### Added
- Added Dutch
- Added Arabic (thanks to [Ahmad Huran](https://github.com/AhmadHuran))
- Started keeping the changelog.

### Changed
- Set language preference in cookie as well as browser storage
- Allow Internation in multi deck environments
- Auto-switch reading direction if the language has `direction: "rtl"` added


## [1.0.4] - 2020-07-01
### Added
- Added possibility to have translatable items outside the .slides
### Changed
- Changes of some formatting.
- Verticator now references the Reveal files by CDN.



## [1.0.3] - 2020-06-25
(Version not published on NPM)
### Added
- Added more demo pages.
### Changed
- Option to change 'langattribute'
- Option to change 'switchselector'



## [1.0.2] - 2020-06-25
### Changed
- Updated readme
- Bugfixes



## [1.0.1] - 2020-06-24
### Changed
- Updated readme
- Changed dropdown to custom



## [1.0.0] - 2020-06-24
- Initial commit

plugin/internation/plugin-src.js
const Plugin = () => {

	Object.filter = (obj, predicate) => 
	Object.keys(obj)
		.filter( key => predicate(obj[key]) )
		.reduce( (res, key) => (res[key] = obj[key], res), {} );
		

	var getParams = function (url) {
		var params = {};
		var parser = document.createElement('a');
		parser.href = url;
		var query = parser.search.substring(1);
		var vars = query.split('&');
		for (var i = 0; i < vars.length; i++) {
			var pair = vars[i].split('=');
			params[pair[0]] = decodeURIComponent(pair[1]);
		}
		return params;
	};

	const getCookie = function (name) {
		let value = `; ${document.cookie}`;
		let parts = value.split(`; ${name}=`);
		if (parts.length === 2) return parts.pop().split(';').shift();
	}


	const readJson = (file, options = { method: 'get' }) =>
		new Promise((resolve, reject) => {
			let request = new XMLHttpRequest();
			request.onload = resolve;
			request.onerror = reject;
			request.overrideMimeType("application/json");
			request.open(options.method, file, true);
			request.onreadystatechange = () => {
				if (request.readyState === 4 && request.status === "200") {
					resolve(request.responseText);
				}
			};
			request.send(null);
		});

		

	function downloadObjectAsJson(exportObj, exportName) {
		let myjson = JSON.stringify(exportObj);
		myjson = myjson.replace(/\\n/g, '').replace(/"\s+|\s+"/g,'"');
		myjson = JSON.stringify((JSON.parse(myjson)), null, 4);
		let dataStr = "data:text/json;charset=utf-8," + encodeURIComponent(myjson); 
		var downloadAnchorNode = document.createElement('a');
		downloadAnchorNode.setAttribute("href", dataStr);
		downloadAnchorNode.setAttribute("download", exportName + ".json");
		document.body.appendChild(downloadAnchorNode); // required for firefox
		downloadAnchorNode.click();
		downloadAnchorNode.remove();
	}

	var escapeJSON = function(str) {
		return str.replace(/\\/g,'\\');
	};

	const InterNation = function(deck, options) {

		let revealEl = deck.getRevealElement();

		const debugLog = function(text) {
			if (options.debug) console.log(text);
		}

		const attributeLooper = function(langattributes, pickdict, origdict, sectionid) {
	
			langattributes.forEach(element => {
	
				if (element.getAttribute(options.langattribute)) {

					let msg = element.getAttribute(options.langattribute);

					if (pickdict[sectionid]) {

						if (pickdict[sectionid][msg]) {
							if (options.html) {
								element.innerHTML = pickdict[sectionid][msg];
							} else {
								element.textContent = pickdict[sectionid][msg];
							}

						}
					} else if (origdict[sectionid]) {
						if (origdict[sectionid][msg]) {
							if (options.html) {
								element.innerHTML = origdict[sectionid][msg];
							} else {
								element.textContent = origdict[sectionid][msg];
							}
						}
					}
				}

			});
			
		}

		const getTextContent = function(deck, options) {

			let sections = revealEl.querySelectorAll(`section`);
			let JSON = false;

			if (langs[options.locale]) {
				if (typeof langs[options.locale].dictionary == "object") {
					debugLog("JSON file is OK and will overrule local textcontent.");
					JSON = true;
				} else {
					debugLog("JSON file is linked but the path is wrong, will parse HTML text.");
					JSON = false;
				}
			} else {
				debugLog("No JSON file for local language, will parse HTML text.");
				langs[options.locale] = {};
				JSON = false;
			}

			if (JSON == false) {
				let newdict = {};
				newdict.slides = {};

				let idDivs = revealEl.querySelectorAll(":scope > [id]");

				idDivs.forEach(idDiv => {
					newdict[idDiv.id] = {}
					let langattributes = idDiv.querySelectorAll(`[${options.langattribute}]`);

					langattributes.forEach(element => {
						let key = element.getAttribute(options.langattribute);
						newdict[idDiv.id][key] = options.html ? escapeJSON(element.innerHTML) : element.textContent;
					});
				});

				sections.forEach(section => {
					if (section.id && !section.classList.contains("stack")) {
						let sectionid = section.id;

						newdict.slides[sectionid] = {}

						let langattributes = section.querySelectorAll(`[${options.langattribute}]`);

						langattributes.forEach(element => {
							let key = element.getAttribute(options.langattribute);
							newdict.slides[sectionid][key] = options.html ? escapeJSON(element.innerHTML) : element.textContent;
						});
					}
				});
				langs[options.locale].name = options.localename;
				langs[options.locale].dictionary = newdict;
			}
			if (options.makejson) {
				let params = getParams(window.location.href);
				if (params.makejson) {
					downloadObjectAsJson(langs[options.locale].dictionary, `${options.locale}`)
				}
			}
		}

		const setText = function(pickLang) {

			let pickdict = langs[pickLang].dictionary;
			let origdict = langs[options.locale].dictionary;

			if (langs[pickLang].direction && langs[pickLang].direction == 'rtl') {
				deck.configure({rtl:true})
			} else {
				deck.configure({rtl:false})
			}

			// Set language for elements outside 'slides'
			let idDivs = revealEl.querySelectorAll(":scope > [id]:not(.slides)");
			idDivs.forEach(idDiv => {
				let divid = idDiv.id;
				let langattributes = idDiv.querySelectorAll(`[${options.langattribute}]`);
				attributeLooper(langattributes, pickdict, origdict, divid);
			});

			// Set language for elements inside 'slides'
			if (pickdict.slides) {
				let sections = revealEl.querySelectorAll(`section[id]`);
				pickdict = pickdict.slides;
				sections.forEach(section => {
					if (!section.classList.contains("stack")) {
						let sectionid = section.id;
						let langattributes = section.querySelectorAll(`[${options.langattribute}]`);
						attributeLooper(langattributes, pickdict, origdict, sectionid);
					}
				});
			}
		}

		const switchSetter = function(selects, value) {
			selects.forEach(thisselect => {
				thisselect.value = value;
				deck.getRevealElement().setAttribute('lang', value);
				let radio = thisselect.querySelector(`input[value='${value}']`);
				if (radio) {
					radio.checked = true;
				}
			});
		}

		const langSwitcher = function() {
			let selects = deck.getRevealElement().querySelectorAll(options.switchselector);

			if (sessionStorage['InternationSettingsStorage'] || getCookie('InternationSettings')) {
				let langPref = sessionStorage['InternationSettingsStorage'] ? sessionStorage['InternationSettingsStorage'] : getCookie('InternationSettings');
				switchSetter(selects, langPref);
				setText(langPref);
			}

			selects.forEach(thisselect => {

				thisselect.addEventListener('change', function(event) {
					switchSetter(selects, event.target.value);
					setText(event.target.value);
					sessionStorage['InternationSettingsStorage'] = event.target.value;
					document.cookie = `InternationSettings=${event.target.value};max-age=${60 * 60 * 24 * 14}`;
				});

			});

		}

		let langattributes = deck.getRevealElement().querySelectorAll(`[${options.langattribute}]`);
		let langs = options.languages;

		if (!langattributes) {
			debugLog(`There are no elements that have the data attribute of ${options.langattribute}`)
		}
		if (Object.keys(langs).length === 0 && langs.constructor === Object) {
			debugLog(`There are no languages defined.`);
			deck.on( 'ready', event => {
				getTextContent(deck, options);
			});

		}

		
		if (langattributes.length > 0) {

			let size = Object.keys(langs).length;
			let counter = 0;

			Object.keys(langs).forEach(function(abbr) {

				if (typeof langs[abbr].dictionary == "object") {

					counter++;

					if (counter == size) {
						getTextContent(deck, options);
						langSwitcher();
					}

				} else {

					readJson(langs[abbr].dictionary).then(res => {
						if (res.srcElement.status != "200") {
							debugLog(`The language file "${langs[abbr].name}" at "${langs[abbr].dictionary}" was not found`);
						} else {
							langs[abbr].dictionary = JSON.parse(res.srcElement.response);
						}

						if (counter == size) {
							getTextContent(deck, options);
							langSwitcher();
						}
					});

					counter++;

				}
			});

			debugLog("Loaded languages:");
			debugLog(langs);

		} else {
			debugLog(`There are no elements that have the data attribute of ${options.langattribute}`)
		}
	}

	const init = function(deck) {

		let defaultOptions = {
			locale: "en",
			localename: 'English',
			langattribute: "data-i18n",
			switchselector: ".langchooser",
			html: true,
			languages: {},
			debug: false,
			makejson: false
		};

		const defaults = function(options, defaultOptions) {
			for (let i in defaultOptions) {
				if (!options.hasOwnProperty(i)) {
					options[i] = defaultOptions[i];
				}
			}
		}

		let options = deck.getConfig().internation || {};
		defaults(options, defaultOptions);

		InterNation(deck, options);
	};

	return {
		id: 'internation',
		init: init
	};
};

export default Plugin;

plugin/internation/internation.js

/*****************************************************************
 * @author: Martijn De Jongh (Martino), martijn.de.jongh@gmail.com
 * https://github.com/Martinomagnifico
 *
 * Internation.js for Reveal.js 
 * Version 1.1.0
 * 
 * @license 
 * MIT licensed
 *
 * Thanks to:
 *  - Hakim El Hattab, Reveal.js 
 ******************************************************************/


(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
  typeof define === 'function' && define.amd ? define(factory) :
  (global = typeof globalThis !== 'undefined' ? globalThis : global || self, global.Internation = factory());
})(this, (function () { 'use strict';

  function _typeof(obj) {
    "@babel/helpers - typeof";

    return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) {
      return typeof obj;
    } : function (obj) {
      return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj;
    }, _typeof(obj);
  }

  var Plugin = function Plugin() {
    Object.filter = function (obj, predicate) {
      return Object.keys(obj).filter(function (key) {
        return predicate(obj[key]);
      }).reduce(function (res, key) {
        return res[key] = obj[key], res;
      }, {});
    };

    var getParams = function getParams(url) {
      var params = {};
      var parser = document.createElement('a');
      parser.href = url;
      var query = parser.search.substring(1);
      var vars = query.split('&');

      for (var i = 0; i < vars.length; i++) {
        var pair = vars[i].split('=');
        params[pair[0]] = decodeURIComponent(pair[1]);
      }

      return params;
    };

    var getCookie = function getCookie(name) {
      var value = "; ".concat(document.cookie);
      var parts = value.split("; ".concat(name, "="));
      if (parts.length === 2) return parts.pop().split(';').shift();
    };

    var readJson = function readJson(file) {
      var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {
        method: 'get'
      };
      return new Promise(function (resolve, reject) {
        var request = new XMLHttpRequest();
        request.onload = resolve;
        request.onerror = reject;
        request.overrideMimeType("application/json");
        request.open(options.method, file, true);

        request.onreadystatechange = function () {
          if (request.readyState === 4 && request.status === "200") {
            resolve(request.responseText);
          }
        };

        request.send(null);
      });
    };

    function downloadObjectAsJson(exportObj, exportName) {
      var myjson = JSON.stringify(exportObj);
      myjson = myjson.replace(/\\n/g, '').replace(/"\s+|\s+"/g, '"');
      myjson = JSON.stringify(JSON.parse(myjson), null, 4);
      var dataStr = "data:text/json;charset=utf-8," + encodeURIComponent(myjson);
      var downloadAnchorNode = document.createElement('a');
      downloadAnchorNode.setAttribute("href", dataStr);
      downloadAnchorNode.setAttribute("download", exportName + ".json");
      document.body.appendChild(downloadAnchorNode); // required for firefox

      downloadAnchorNode.click();
      downloadAnchorNode.remove();
    }

    var escapeJSON = function escapeJSON(str) {
      return str.replace(/\\/g, '\\');
    };

    var InterNation = function InterNation(deck, options) {
      var revealEl = deck.getRevealElement();

      var debugLog = function debugLog(text) {
        if (options.debug) console.log(text);
      };

      var attributeLooper = function attributeLooper(langattributes, pickdict, origdict, sectionid) {
        langattributes.forEach(function (element) {
          if (element.getAttribute(options.langattribute)) {
            var msg = element.getAttribute(options.langattribute);

            if (pickdict[sectionid]) {
              if (pickdict[sectionid][msg]) {
                if (options.html) {
                  element.innerHTML = pickdict[sectionid][msg];
                } else {
                  element.textContent = pickdict[sectionid][msg];
                }
              }
            } else if (origdict[sectionid]) {
              if (origdict[sectionid][msg]) {
                if (options.html) {
                  element.innerHTML = origdict[sectionid][msg];
                } else {
                  element.textContent = origdict[sectionid][msg];
                }
              }
            }
          }
        });
      };

      var getTextContent = function getTextContent(deck, options) {
        var sections = revealEl.querySelectorAll("section");
        var JSON = false;

        if (langs[options.locale]) {
          if (_typeof(langs[options.locale].dictionary) == "object") {
            debugLog("JSON file is OK and will overrule local textcontent.");
            JSON = true;
          } else {
            debugLog("JSON file is linked but the path is wrong, will parse HTML text.");
            JSON = false;
          }
        } else {
          debugLog("No JSON file for local language, will parse HTML text.");
          langs[options.locale] = {};
          JSON = false;
        }

        if (JSON == false) {
          var newdict = {};
          newdict.slides = {};
          var idDivs = revealEl.querySelectorAll(":scope > [id]");
          idDivs.forEach(function (idDiv) {
            newdict[idDiv.id] = {};
            var langattributes = idDiv.querySelectorAll("[".concat(options.langattribute, "]"));
            langattributes.forEach(function (element) {
              var key = element.getAttribute(options.langattribute);
              newdict[idDiv.id][key] = options.html ? escapeJSON(element.innerHTML) : element.textContent;
            });
          });
          sections.forEach(function (section) {
            if (section.id && !section.classList.contains("stack")) {
              var sectionid = section.id;
              newdict.slides[sectionid] = {};

              var _langattributes = section.querySelectorAll("[".concat(options.langattribute, "]"));

              _langattributes.forEach(function (element) {
                var key = element.getAttribute(options.langattribute);
                newdict.slides[sectionid][key] = options.html ? escapeJSON(element.innerHTML) : element.textContent;
              });
            }
          });
          langs[options.locale].name = options.localename;
          langs[options.locale].dictionary = newdict;
        }

        if (options.makejson) {
          var params = getParams(window.location.href);

          if (params.makejson) {
            downloadObjectAsJson(langs[options.locale].dictionary, "".concat(options.locale));
          }
        }
      };

      var setText = function setText(pickLang) {
        var pickdict = langs[pickLang].dictionary;
        var origdict = langs[options.locale].dictionary;

        if (langs[pickLang].direction && langs[pickLang].direction == 'rtl') {
          deck.configure({
            rtl: true
          });
        } else {
          deck.configure({
            rtl: false
          });
        } // Set language for elements outside 'slides'


        var idDivs = revealEl.querySelectorAll(":scope > [id]:not(.slides)");
        idDivs.forEach(function (idDiv) {
          var divid = idDiv.id;
          var langattributes = idDiv.querySelectorAll("[".concat(options.langattribute, "]"));
          attributeLooper(langattributes, pickdict, origdict, divid);
        }); // Set language for elements inside 'slides'

        if (pickdict.slides) {
          var sections = revealEl.querySelectorAll("section[id]");
          pickdict = pickdict.slides;
          sections.forEach(function (section) {
            if (!section.classList.contains("stack")) {
              var sectionid = section.id;

              var _langattributes2 = section.querySelectorAll("[".concat(options.langattribute, "]"));

              attributeLooper(_langattributes2, pickdict, origdict, sectionid);
            }
          });
        }
      };

      var switchSetter = function switchSetter(selects, value) {
        selects.forEach(function (thisselect) {
          thisselect.value = value;
          deck.getRevealElement().setAttribute('lang', value);
          var radio = thisselect.querySelector("input[value='".concat(value, "']"));

          if (radio) {
            radio.checked = true;
          }
        });
      };

      var langSwitcher = function langSwitcher() {
        var selects = deck.getRevealElement().querySelectorAll(options.switchselector);

        if (sessionStorage['InternationSettingsStorage'] || getCookie('InternationSettings')) {
          var langPref = sessionStorage['InternationSettingsStorage'] ? sessionStorage['InternationSettingsStorage'] : getCookie('InternationSettings');
          switchSetter(selects, langPref);
          setText(langPref);
        }

        selects.forEach(function (thisselect) {
          thisselect.addEventListener('change', function (event) {
            switchSetter(selects, event.target.value);
            setText(event.target.value);
            sessionStorage['InternationSettingsStorage'] = event.target.value;
            document.cookie = "InternationSettings=".concat(event.target.value, ";max-age=").concat(60 * 60 * 24 * 14);
          });
        });
      };

      var langattributes = deck.getRevealElement().querySelectorAll("[".concat(options.langattribute, "]"));
      var langs = options.languages;

      if (!langattributes) {
        debugLog("There are no elements that have the data attribute of ".concat(options.langattribute));
      }

      if (Object.keys(langs).length === 0 && langs.constructor === Object) {
        debugLog("There are no languages defined.");
        deck.on('ready', function (event) {
          getTextContent(deck, options);
        });
      }

      if (langattributes.length > 0) {
        var size = Object.keys(langs).length;
        var counter = 0;
        Object.keys(langs).forEach(function (abbr) {
          if (_typeof(langs[abbr].dictionary) == "object") {
            counter++;

            if (counter == size) {
              getTextContent(deck, options);
              langSwitcher();
            }
          } else {
            readJson(langs[abbr].dictionary).then(function (res) {
              if (res.srcElement.status != "200") {
                debugLog("The language file \"".concat(langs[abbr].name, "\" at \"").concat(langs[abbr].dictionary, "\" was not found"));
              } else {
                langs[abbr].dictionary = JSON.parse(res.srcElement.response);
              }

              if (counter == size) {
                getTextContent(deck, options);
                langSwitcher();
              }
            });
            counter++;
          }
        });
        debugLog("Loaded languages:");
        debugLog(langs);
      } else {
        debugLog("There are no elements that have the data attribute of ".concat(options.langattribute));
      }
    };

    var init = function init(deck) {
      var defaultOptions = {
        locale: "en",
        localename: 'English',
        langattribute: "data-i18n",
        switchselector: ".langchooser",
        html: true,
        languages: {},
        debug: false,
        makejson: false
      };

      var defaults = function defaults(options, defaultOptions) {
        for (var i in defaultOptions) {
          if (!options.hasOwnProperty(i)) {
            options[i] = defaultOptions[i];
          }
        }
      };

      var options = deck.getConfig().internation || {};
      defaults(options, defaultOptions);
      InterNation(deck, options);
    };

    return {
      id: 'internation',
      init: init
    };
  };

  return Plugin;

}));
